CefSharp/Internals/AsyncTaskCompletionSource.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright © 2010-2017 The CefSharp Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CefSharp
+{
+    /// <summary>
+    /// Wraps <see cref="TaskCompletionSource{T}"/> by providing a way to set its result in an async fashion.
+    /// This prevents the Task Continuation being executed sync on the same thread, which may be required when
+    /// continuations must not run on CEF UI threads.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class AsyncTaskCompletionSource<T>
```

Comment:
This can probably be made internal.

Reply:
Last I checked IntetnalsVisibleTo was working for all the c# projects, just doesn't work with the VC++ ones, so we do have to leave it public if it's going to be used in C++. 

https://github.com/cefsharp/CefSharp/blob/master/CefSharp/Properties/AssemblyInfo.cs#L26

Reply:
Oh sorry, I missed the InternalsVisibleTo. Well then it's ok to make it internal except if we also want to replace the two usages of TaskExtensions::TrySetResultAsync which are in the C++ project.

Reply:
Ok no worries, will have to remain public.

Reply:
guys why dont u use the default TaskCompletionSource and create it with TaskCreationOptions.RunContinuationsAsync ?

Reply:
> guys why dont u use the default TaskCompletionSource and create it with TaskCreationOptions.RunContinuationsAsync ?

TaskCreationOptions.RunContinuationsAsync was added in .Net 4.6

We are still supporting .Net 4.5.2 (no plans to change this)

One day I'll get around to updating the .Net Core builds to use RunContinuationsAsync